test/IISIntegration.FunctionalTests/Inprocess/AppOfflineTests.cs

```diff
@@ -0,0 +1,108 @@
+        {
+            var deploymentParameters = Helpers.GetBaseDeploymentParameters();
+
+            var deploymentResult = await DeployAsync(deploymentParameters);
+
+            AddAppOffline(deploymentResult.DeploymentResult.ContentRoot);
+
+            var response = await deploymentResult.RetryingHttpClient.GetAsync("HelloWorld");
+
+            Assert.Equal(HttpStatusCode.ServiceUnavailable, response.StatusCode);
+        }
+
+        [ConditionalFact]
+        public async Task AppOfflineDroppedWhileSiteRunning_SiteShutsDown()
+        {
+            var deploymentResult = await StartSite();
+
+            AddAppOffline(deploymentResult.DeploymentResult.ContentRoot);
+            var response = await deploymentResult.RetryingHttpClient.GetAsync("HelloWorld");
+
+            while (response.IsSuccessStatusCode)
+            {
+                // Keep retrying until app_offline is present.
+                response = await deploymentResult.RetryingHttpClient.GetAsync("HelloWorld");
+            }
+
+            Assert.Equal(HttpStatusCode.ServiceUnavailable, response.StatusCode);
+
+            deploymentResult.DeploymentResult.HostShutdownToken.WaitHandle.WaitOne(millisecondsTimeout: 1000);
+
+            Assert.True(deploymentResult.DeploymentResult.HostShutdownToken.IsCancellationRequested);
+        }
+
+        [ConditionalFact]
+        public async Task AppOfflineDropped_CanRemoveAppOfflineAfterAdding()
+        {
+            var deploymentResult = await StartSite();
+
+            AddAppOffline(deploymentResult.DeploymentResult.ContentRoot);
+
+            var response = await deploymentResult.RetryingHttpClient.GetAsync("HelloWorld");
+
+            while (response.IsSuccessStatusCode)
+            {
+                // Keep retrying until app_offline is present.
+                response = await deploymentResult.RetryingHttpClient.GetAsync("HelloWorld");
+            }
+
+            Assert.Equal(HttpStatusCode.ServiceUnavailable, response.StatusCode);
+
+            RemoveAppOffline(deploymentResult.DeploymentResult.ContentRoot);
+        }
+
+        private async Task<IISDeploymentResult> StartSite()
+        {
+            var deploymentParameters = Helpers.GetBaseDeploymentParameters();
+
+            var deploymentResult = await DeployAsync(deploymentParameters);
+
+            var response = await deploymentResult.RetryingHttpClient.GetAsync("HelloWorld");
+
+            var responseText = await response.Content.ReadAsStringAsync();
+            Assert.Equal("Hello World", responseText);
+            return deploymentResult;
+        }
+
+        private void AddAppOffline(string appPath)
+        {
+            var fs = File.Create(Path.Combine(appPath, "app_offline.htm"));
+            // Close the file to allow for 
+            fs.Close();
+        }
+
+        private void RemoveAppOffline(string appPath)
+        {
+
+            RetryHelper.RetryOperation(
```

Comment:
Do we lock appofline?

Reply:
I think we do to read the contents of the file.

Reply:
Right, but we should set FILE_SHARE_READ | WRITE | DELETE while doing so.